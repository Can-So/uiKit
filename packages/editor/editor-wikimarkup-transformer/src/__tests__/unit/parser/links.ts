import WikiMarkupTransformer from '../../../index';

describe('JIRA wiki markup - Links', () => {
  const testCases: Array<[string, string]> = [
    [
      'should convert simple links to link marks',
      'This is a [https://www.atlassian.com] link',
    ],
    [
      'should convert links with text to link marks',
      'This is an [Atlassian|https://www.atlassian.com] link',
    ],
    [
      'should convert mailto: links to link marks + remove mailto: prefix',
      'This is a [mailto:legendaryservice@atlassian.com] link',
    ],
    [
      'should ignore file links',
      'This is a local file [file:///c:/temp/foo.txt] link',
    ],
    ['should ignore anchor-links starting with #', 'This is an [#anchor] link'],
    ['should ignore anchors', 'Anchor {anchor:anchorname} here'],
    [
      'should convert links with port to link marks',
      'Link with port [http://localhost:8080] here',
    ],
    [
      'should convert normal url text to link marks',
      'Hey, checkout atlassian=http://www.atlassian.com',
    ],
    [
      'should convert url text inside <> to link marks',
      'Hey, this is amazing <https://www.atlassian.com>',
    ],
    [
      'should convert IRC address text',
      'An IRC address: irc://atlassian.com/confluence',
    ],
    [
      'should ignore invalid protocal',
      'An invalid URL: invalid://nowhere.com/hello',
    ],
    [
      'should create Media Image and Link from linkable-image',
      'This is a linkable image [!image.jpg!|https://www.atlassian.com]',
    ],
    [
      '[CS-240] should know where to end the link',
      `[Link Title|http://www.google.com] boy I hope this doesn't go all the way to here] that would be bad.`,
    ],
    [
      '[CS-385] should link text in link format',
      `[https://splunk.paas-inf.net/en-GB/app/search/search?earliest=\-1d&latest=now|https://www.google.com]`,
    ],
  ];

  for (const [testCaseDescription, markup] of testCases) {
    fit(testCaseDescription, () => {
      const transformer = new WikiMarkupTransformer();
      expect(transformer.parse(markup)).toMatchSnapshot();
    });
  }
});
