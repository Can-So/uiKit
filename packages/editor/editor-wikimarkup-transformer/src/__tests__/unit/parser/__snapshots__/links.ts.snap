// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JIRA wiki markup - Links [CS-240] should know where to end the link 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "marks": Array [
            Object {
              "attrs": Object {
                "__confluenceMetadata": null,
                "href": "http://www.google.com",
              },
              "type": "link",
            },
          ],
          "text": "Link Title",
          "type": "text",
        },
        Object {
          "text": " boy I hope this doesn't go all the way to here] that would be bad.",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`JIRA wiki markup - Links [CS-385] should link text in link format 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "marks": Array [
            Object {
              "attrs": Object {
                "__confluenceMetadata": null,
                "href": "https://www.google.com",
              },
              "type": "link",
            },
          ],
          "text": "https://splunk.paas-inf.net/en-GB/app/search/search?earliest=-1d&latest=now",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`JIRA wiki markup - Links [CS-478] should resolve link with | in the url 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "marks": Array [
            Object {
              "attrs": Object {
                "__confluenceMetadata": null,
                "href": "https://hello.atlassian.net/wiki/spaces/Engage/pages/296780133/EP+Chrome+Extension#Set-test-metadata-without-having-to-create-a-message-in-targeting--|-title-=-3rd-Iteration-|-colour-=-Red-|-MAKEITSO-3652",
              },
              "type": "link",
            },
          ],
          "text": "page",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`JIRA wiki markup - Links should convert IRC address text 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "An IRC address: ",
          "type": "text",
        },
        Object {
          "marks": Array [
            Object {
              "attrs": Object {
                "__confluenceMetadata": null,
                "href": "irc://atlassian.com/confluence",
              },
              "type": "link",
            },
          ],
          "text": "irc://atlassian.com/confluence",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`JIRA wiki markup - Links should convert links with port to link marks 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "Link with port ",
          "type": "text",
        },
        Object {
          "marks": Array [
            Object {
              "attrs": Object {
                "__confluenceMetadata": null,
                "href": "http://localhost:8080",
              },
              "type": "link",
            },
          ],
          "text": "http://localhost:8080",
          "type": "text",
        },
        Object {
          "text": " here",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`JIRA wiki markup - Links should convert links with text to link marks 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "This is an ",
          "type": "text",
        },
        Object {
          "marks": Array [
            Object {
              "attrs": Object {
                "__confluenceMetadata": null,
                "href": "https://www.atlassian.com",
              },
              "type": "link",
            },
          ],
          "text": "Atlassian",
          "type": "text",
        },
        Object {
          "text": " link",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`JIRA wiki markup - Links should convert mailto: links to link marks + remove mailto: prefix 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "This is a ",
          "type": "text",
        },
        Object {
          "marks": Array [
            Object {
              "attrs": Object {
                "__confluenceMetadata": null,
                "href": "mailto:legendaryservice@atlassian.com",
              },
              "type": "link",
            },
          ],
          "text": "legendaryservice@atlassian.com",
          "type": "text",
        },
        Object {
          "text": " link",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`JIRA wiki markup - Links should convert normal url text to link marks 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "Hey, checkout atlassian=",
          "type": "text",
        },
        Object {
          "marks": Array [
            Object {
              "attrs": Object {
                "__confluenceMetadata": null,
                "href": "http://www.atlassian.com",
              },
              "type": "link",
            },
          ],
          "text": "http://www.atlassian.com",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`JIRA wiki markup - Links should convert simple links to link marks 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "This is a ",
          "type": "text",
        },
        Object {
          "marks": Array [
            Object {
              "attrs": Object {
                "__confluenceMetadata": null,
                "href": "https://www.atlassian.com",
              },
              "type": "link",
            },
          ],
          "text": "https://www.atlassian.com",
          "type": "text",
        },
        Object {
          "text": " link",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`JIRA wiki markup - Links should convert url text inside <> to link marks 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "Hey, this is amazing <",
          "type": "text",
        },
        Object {
          "marks": Array [
            Object {
              "attrs": Object {
                "__confluenceMetadata": null,
                "href": "https://www.atlassian.com",
              },
              "type": "link",
            },
          ],
          "text": "https://www.atlassian.com",
          "type": "text",
        },
        Object {
          "text": ">",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`JIRA wiki markup - Links should create Media Image and Link from linkable-image 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "This is a linkable image ",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "attrs": Object {
        "layout": "center",
        "width": null,
      },
      "content": Array [
        Object {
          "attrs": Object {
            "__displayType": null,
            "__fileMimeType": null,
            "__fileName": null,
            "__fileSize": null,
            "__key": null,
            "collection": "",
            "height": 183,
            "id": "image.jpg",
            "occurrenceKey": null,
            "type": "file",
            "url": null,
            "width": 200,
          },
          "type": "media",
        },
      ],
      "type": "mediaSingle",
    },
    Object {
      "content": Array [
        Object {
          "marks": Array [
            Object {
              "attrs": Object {
                "__confluenceMetadata": null,
                "href": "https://www.atlassian.com",
              },
              "type": "link",
            },
          ],
          "text": "!image.jpg!",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`JIRA wiki markup - Links should ignore anchor-links starting with # 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "This is an [#anchor] link",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`JIRA wiki markup - Links should ignore anchors 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "Anchor  here",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`JIRA wiki markup - Links should ignore file links 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "This is a local file [file:///c:/temp/foo.txt] link",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`JIRA wiki markup - Links should ignore invalid protocal 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "An invalid URL: invalid://nowhere.com/hello",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;
